**Overview**
Short summary of project: The overall goal of the project is to find the target sites where the insertion sequence (IS) transposases into. The IS transposes itself into genomes in a sequence specific manner.

**Dataset details**
Sequencing platform: Use publicly available IS data from https://isfinder.biotoul.fr/

Sample size: 350 sequences
Experimental groups: One IS family (IS110), with one sub-group (IS1111)
Species: Bacteria

Compute resources
Compute platform: Gadi and Artemis
Gadi shortcode: er01. BLAST against NCBI non-redundant nucleotide database and insertion site flank extraction performed on NCI Gadi HPC, using project code er01
DashR project name: SIHsandbox. Downstream analysis performed on USyd Artemis HPC using SIHsandbox project code

Code written by: Rezwan Siddiquee and Cali Willet (SIH).

Acknowledgements
The authors acknowledge the technical assistance provided by the Sydney Informatics Hub, a Core Research Facility of the University of Sydney and the Australian BioCommons which is enabled by NCRIS via Bioplatforms Australia.

The authors acknowledge the use of the National Computational Infrastructure (NCI) supported by the Australian Government and the Sydney Informatics Hub HPC Allocation Scheme, supported by the Deputy Vice-Chancellor (Research), University of Sydney and the ARC LIEF, 2019: Smith, Muller, Thornber et al., Sustaining and strengthening merit-based access to National Computational Infrastructure (LE190100021).

For USyd's Artemis HPC: The authors acknowledge the technical assistance provided by the Sydney Informatics Hub, a Core Research Facility of the University of Sydney and the Australian BioCommons which is enabled by NCRIS via Bioplatforms Australia. This research utilised the high performance computing service, Artemis, provided by the Sydney Informatics Hub, Core Research Facility, University of Sydney.
